File: jhello_panama/Hello.java
package jhello_panama;

import java.lang.foreign.*;
import java.lang.management.CompilationMXBean;
import java.lang.management.ManagementFactory;
import java.lang.invoke.MethodHandle;
import java.nio.file.Path;

public final class Hello
{
    private static final long WARMUP_NANOS = 10_000_000_000L;
    private static final int WARMUP_COUNT = 1_000_000;
    private static final int STABLE_JIT_CHECKS = 3;

    private static final Linker linker = Linker.nativeLinker();
    private static final SymbolLookup symbolLookup;
    private static final MethodHandle plusone;
    private static final MethodHandle currentTimestamp;
    
    static {
        try {
            // Load the shared library
            java.io.File currentDir = new java.io.File(".").getCanonicalFile();
            System.load(currentDir.getPath() + "/newplus/libnewplus.so");
            symbolLookup = SymbolLookup.loaderLookup();
            
            // Define function signatures
            FunctionDescriptor plusoneDesc = FunctionDescriptor.of(ValueLayout.JAVA_INT, ValueLayout.JAVA_INT);
            FunctionDescriptor timestampDesc = FunctionDescriptor.of(ValueLayout.JAVA_LONG);
            
            // Get method handles
            plusone = linker.downcallHandle(
                symbolLookup.find("plusone").orElseThrow(),
                plusoneDesc
            );
            
            currentTimestamp = linker.downcallHandle(
                symbolLookup.find("current_timestamp").orElseThrow(),
                timestampDesc
            );
        } catch (Exception e) {
            throw new RuntimeException("Failed to initialize native bindings", e);
        }
    }
    
    public static int plusone(int x) {
        try {
            return (int) plusone.invokeExact(x);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }
    
    public static long current_timestamp() {
        try {
            return (long) currentTimestamp.invokeExact();
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }
    
    static int runLoop(int count) {
        int x = 0;
        while (x < count)
            x = plusone(x);

        return x;
    }

    static void warmup() {
        final CompilationMXBean compiler = ManagementFactory.getCompilationMXBean();
        final long start = System.nanoTime();
        long loadedClasses = ManagementFactory.getClassLoadingMXBean().getTotalLoadedClassCount();
        long compilationTime = compiler.getTotalCompilationTime();
        int stableChecks = 0;

        do {
            runLoop(WARMUP_COUNT);

            final long newLoadedClasses = ManagementFactory.getClassLoadingMXBean().getTotalLoadedClassCount();
            final long newCompilationTime = compiler.getTotalCompilationTime();
            if (loadedClasses == newLoadedClasses && compilationTime == newCompilationTime)
                stableChecks++;
            else
                stableChecks = 0;

            loadedClasses = newLoadedClasses;
            compilationTime = newCompilationTime;
        } while (System.nanoTime() - start < WARMUP_NANOS || stableChecks < STABLE_JIT_CHECKS);
    }

    static void run(int count) {
        final long start = current_timestamp();
        runLoop(count);
        System.out.println(current_timestamp() - start);
    }
    
    public static void main(String[] args) throws Exception {
        if (args.length == 0) {
            System.err.println("First arg (0 - 2000000000) is required.");
            return;
        }
        
        int count = Integer.parseInt(args[0]);
        if (count <= 0 || count > 2000000000) {
            System.err.println("Must be a positive number not exceeding 2 billion.");
            return;
        }
        
        // Warmup
        System.gc();
        warmup();
        
        // Run benchmark
        run(count);
    }
}

File: clojure_panama/hello.clj
(ns clojure-panama.hello
  (:gen-class)
  (:require
   [criterium.core :as crit])
  (:import
   [java.lang.foreign Arena FunctionDescriptor Linker Linker$Option MemoryLayout SymbolLookup ValueLayout]
   [java.lang.invoke MethodHandle MethodHandleProxies]
   [java.util Optional]
   [java.util.function IntUnaryOperator LongSupplier]))

(set! *warn-on-reflection* true)

(def ^Linker linker (Linker/nativeLinker))
(def ^SymbolLookup symbol-lookup
  (SymbolLookup/libraryLookup
   (.toPath (.getCanonicalFile (java.io.File. "../newplus/libnewplus.so")))
   (Arena/global)))
(def linker-options (make-array Linker$Option 0))

(defn downcall-handle [symbol descriptor]
  (.downcallHandle
   linker
   (.orElseThrow ^Optional (.find symbol-lookup symbol))
   descriptor
   linker-options))

(def ^IntUnaryOperator plusone
  (MethodHandleProxies/asInterfaceInstance
   IntUnaryOperator
   ^MethodHandle
   (downcall-handle
    "plusone"
    (FunctionDescriptor/of
     ValueLayout/JAVA_INT
     (into-array MemoryLayout [ValueLayout/JAVA_INT])))))

(def ^LongSupplier current-timestamp
  (MethodHandleProxies/asInterfaceInstance
   LongSupplier
   ^MethodHandle
   (downcall-handle
    "current_timestamp"
    (FunctionDescriptor/of
     ValueLayout/JAVA_LONG
     (make-array MemoryLayout 0)))))

(defn parse-count [args]
  (when-not (= 1 (count args))
    (throw (ex-info "Count must be specified once." {})))
  (let [count (try
                (Long/parseLong (first args))
                (catch NumberFormatException _
                  nil))]
    (when-not (and count (pos? count) (<= count 2000000000))
      (throw (ex-info "Count must be one positive integer not exceeding 2 billion." {})))
    count))

(defn run-benchmark [count print?]
  (let [start (.getAsLong current-timestamp)
        res (do
              (loop [x 0]
                (if (< x count)
                  (recur (.applyAsInt plusone (int x)))
                  nil))
              (- (.getAsLong current-timestamp) start))]
    (when print?
      (println res))))

(defmacro warmup [expr]
  `(crit/warmup-for-jit (* 10 crit/s-to-ns) (fn [] ~expr)))

(defn -main [& args]
  (let [count (parse-count args)]
    (crit/force-gc)
    (warmup (run-benchmark 1000000 false))
    (run-benchmark count true)))
